nuxeo-runtime/nuxeo-runtime-stream/src/main/java/org/nuxeo/runtime/stream/StreamPubSubProvider.java

```diff
@@ -0,0 +1,130 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.function.BiConsumer;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.lib.stream.computation.Record;
+import org.nuxeo.lib.stream.log.LogAppender;
+import org.nuxeo.lib.stream.log.LogRecord;
+import org.nuxeo.lib.stream.log.LogTailer;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.pubsub.AbstractPubSubProvider;
+
+/**
+ * @since 10.1
+ */
+public class StreamPubSubProvider extends AbstractPubSubProvider {
+    private static final Log log = LogFactory.getLog(StreamPubSubProvider.class);
+
+    public static final String GROUP_PREFIX = "pub-sub";
+
+    public static final String NODE_ID_PROP = "repository.clustering.id";
+
+    protected static final Random RANDOM = new Random();
+
+    protected String logConfig;
+
+    protected String logName;
+
+    protected LogAppender<Record> appender;
+
+    protected Thread thread;
+
+    @Override
+    public void initialize(Map<String, String> options, Map<String, List<BiConsumer<String, byte[]>>> subscribers) {
+        log.debug("Initializing ");
+        super.initialize(options, subscribers);
+        logConfig = options.getOrDefault("logConfig", "default");
+        logName = options.get("logName");
+        if (logName == null || logName.isEmpty()) {
+            throw new IllegalArgumentException("Missing option logName in StreamPubSubProviderDescriptor");
+        }
+        appender = Framework.getService(StreamService.class).getLogManager(logConfig).getAppender(logName);
+        startConsumerThread();
+        log.debug("Initialized");
+    }
+
+    protected void startConsumerThread() {
+        Subscriber subscriber = new Subscriber();
+        thread = new Thread(subscriber, "Nuxeo-PubSub-");
+        thread.setUncaughtExceptionHandler((t, e) -> log.error("Uncaught error on thread " + t.getName(), e));
+        thread.setPriority(Thread.NORM_PRIORITY);
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @Override
+    public void publish(String topic, byte[] message) {
+        appender.append(topic, Record.of(topic, message));
+    }
+
+    @Override
+    public void close() {
+        appender = null;
+        if (thread != null) {
+            thread.interrupt();
+            thread = null;
+            log.debug("Closed");
+        }
+    }
+
+    public class Subscriber implements Runnable {
+
+        @Override
+        public void run() {
+            // using different group name enable fan out
+            String group = GROUP_PREFIX + ":" + getNodeId();
+            log.debug("Starting subscriber thread with group: " + group);
+            LogTailer<Record> tailer = Framework.getService(StreamService.class).getLogManager(logConfig).createTailer(
+                    group, logName);
+            // Only interested in new messages
+            tailer.toEnd();
+            while (true) {
```

Comment:
FYI I prefer `for (;;)`, which does the same thing but doesn't contain a literal (true) and can be read as "forever".

Reply:
my eyes read "while true" as forever